### PR TITLE
ENH: Add Operator.norm, closes #1065

### LINKED ISSUE #1065: Add Operator.norm
Computing operator norms happens all the time in code, and we mostly refer to the `odl.power_method_opnorm` method, which is fine but error prone and occationally slow.

I propose that we add a new method called `Operator.norm` with the following default implementation:

```python
def norm(self, estimate=False):
    if not estimate:
        raise NotImplementedError('Operator norm not implemented')
    else:
        return odl.power_method_opnorm(self)
```

Some classes could then over-ride this method when there are known formulae. For example, in `DiagonalOperator` we could implement

```python
def norm(self, estimate=False):
    # Pen-and-paper exercise shows that this holds
    norms = [op.norm(estimate=estimate) for op in self.operators]
    return max(norms)
```

for some other operators I'm sure we could obtain similar bounds.

One obvious issue with this is non-linear operators, so we'd need to solve #551.

### PR DESCRIPTION
**Pull request by adler-j**
Making a small initial PR to review the concept. If we go further we can start adding this to various classes.

### CODE REVIEW

**Comment by adler-j** [comment]
Perhaps @MatthiasJE could give a review of this as well?

File: odl/operator/operator.py [hunk 1]
```diff
@@ -707,6 +707,54 @@ def __call__(self, x, out=None, **kwargs):
...
+        >>> grad = odl.Gradient(spc)
+        >>> opnorm = grad.norm(estimate=True)
+        """
+        if not estimate:
+            raise NotImplementedError('`Operator.norm()` not implemented, use '
+                                      '`Operator.norm(estimate=True)` to '
+                                      'obtain an estimate.')
+        else:
+            from odl.operator.oputils import power_method_opnorm
+            return power_method_opnorm(self)
```
Comment (adler-j) [inline odl/operator/operator.py]: missing **kwargs

File: odl/operator/operator.py [hunk 2]
```diff
@@ -707,6 +707,54 @@ def __call__(self, x, out=None, **kwargs):
                     raise_from(new_exc, err)
         return out
 
+    def norm(self, estimate=False, **kwargs):
+        """Return the operator norm of this operator.
+
+        If this operator is non-linera, this should be the Lipschitz constant.
+
```
Comment (MatthiasJE) [inline odl/operator/operator.py]: typo "lineara"

File: odl/operator/operator.py [hunk 3]
```diff
@@ -707,6 +707,54 @@ def __call__(self, x, out=None, **kwargs):
                     raise_from(new_exc, err)
         return out
 
+    def norm(self, estimate=False, **kwargs):
```
Comment (MatthiasJE) [inline odl/operator/operator.py]: I think it would be good if it was possible to set the operator norm manually. The current framework would not allow for that.

File: odl/operator/operator.py [hunk 4]
```diff
@@ -707,6 +707,54 @@ def __call__(self, x, out=None, **kwargs):
...
+
+        For others, there is no closed form expression and an estimate is
+        needed:
+
+        >>> spc = odl.uniform_discr(0, 1, 3)
+        >>> grad = odl.Gradient(spc)
+        >>> opnorm = grad.norm(estimate=True)
+        """
+        if not estimate:
+            raise NotImplementedError('`Operator.norm()` not implemented, use '
```
Comment (MatthiasJE) [inline odl/operator/operator.py]: Most gradient operators have closed form norms that depend on the discretization. It may be useful to utilize this here.

File: odl/operator/operator.py (hunk 3 above)
Comment (adler-j) [inline odl/operator/operator.py]: Quite skeptical to a global way of doing this. The main reason is that `Operator`s are immutable through-out ODL, which means that if I am given an operator, I know it will always return the same thing at all times, and I don't have to be afraid of it changing values due to some change. While we really only depend on this property for `__call__`, it is health to try to keep it that way (simply because code becomes so much easier to debug).

Even so, this framework allows you to manually set the norm in sub-classes, its just that we don't provide a global way of doing it.

Is there any specific use-case where you would want to set it manually? Perhaps that would give some more insight into how to design this.

File: odl/operator/operator.py (hunk 4 above)
Comment (adler-j) [inline odl/operator/operator.py]: I agree, but that will be added in another PR, in order to keep this simple. Other operators that are prime candidates include the Fourier transform and \*caugh @kohr-h\* convolution operator, as well as `ProductSpaceOperator`.

File: odl/operator/operator.py (hunk 3 above)
Comment (MatthiasJE) [inline odl/operator/operator.py]: At the moment I am looking at X-ray transform examples (e.g. PET, CT). Here you need to estimate the operator norm but depending on the computational cost, you may want to do it differently for each case. For example, on a small scanner you may want to run 100 power iterations, on a big one you may want to do it for 20 iterations and add a 5 % margin to it or to load stored values for it. Of course, this can be done for subclasses that really need it.

Comment (kohr-h) [inline odl/operator/operator.py]: For that case I think you can just steer via `kwargs` how many iterations you run and then add the safety margin yourself.
And yes, keeping things immutable as far as possible is favorable, also in the light of serialization (#1028).

**Comment by kohr-h** [comment]
> Also the changes do not include automatic derivation of norms for composition of operators.

That would be useful, agreed, but then you get two different variants of "estimate" namely "estimate numerically" or "estimate from the norms of the constituents". If we default to the latter, I think it wouldn't be right to call it `norm` anymore since it's only an upper bound.

> reuse computed results.

One way of doing that, and I suggest that we think about that separately from this PR, is to add a cache to this method. A possible implementation of that would be to store a tuple `(num_powmethod_iter, norm_approx)`, and when you call the function with `num_iter <= num_powmethod_iter` the method would return the cached `norm_approx`, and otherwise rerun the iteration and store the new value. A kind of stripped-down version of an LRU cache.
